server/handler/upgradehandler.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package handler
+
+import (
+	"net/http"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/session"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// ClusterUpgradeHandler is the handler for upgrading cluster.
+type ClusterUpgradeHandler struct {
+	store kv.Storage
+}
+
+// NewClusterUpgradeHandler creates a new ClusterUpgradeHandler.
+func NewClusterUpgradeHandler(store kv.Storage) *ClusterUpgradeHandler {
+	return &ClusterUpgradeHandler{store: store}
+}
+
+// ServeHTTP handles request of ddl server info.
+func (h ClusterUpgradeHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		WriteError(w, errors.Errorf("This API only support POST method"))
+		return
+	}
+
+	var err error
+	var hasDone bool
+	op := req.FormValue("op")
+	switch op {
+	case "start":
+		hasDone, err = h.startUpgrade()
+	case "finish":
+		hasDone, err = h.finishUpgrade()
+	default:
+		WriteError(w, errors.Errorf("wrong operation:%s", op))
+		return
+	}
+
+	if err != nil {
+		WriteError(w, err)
+		return
+	}
+	if hasDone {
+		switch op {
+		case "start":
+			WriteData(w, "Be upgrading.")
+		case "finish":
+			WriteData(w, "Be normal.")
+		}
+	}
+	WriteData(w, "success!")
+	logutil.Logger(req.Context()).Info("[upgrading] upgrade op success",
```

Comment:
Please use category(`zap.String("category", "upgrade")`) instead of `[upgrade]`.